modules/amm-executor/src/main/scala/org/ergoplatform/dex/executor/amm/interpreters/v1/N2TCFMMInterpreter.scala

```diff
@@ -0,0 +1,254 @@
+        val predictedDexOutput = Output.fromErgoBox(tx.outputs(2))
+        (tx, nextPool, predictedDexOutput)
+      }
+
+  def swap(
+    swap: SwapErg,
+    pool: CFMMPool
+  ): F[(ErgoLikeTransaction, Traced[Predicted[CFMMPool]], Output)] =
+    resolver.getLatest
+      .flatMap(_.orRaise[F](EmptyOutputForDexTokenFee(pool.poolId, swap.box.boxId)))
+      .flatMap { dexFeeOutput =>
+        swapParamsErgFee(swap, pool).toRaise.flatMap { case (input, output, dexFee) =>
+          (swap match {
+            case s: SwapP2Pk =>
+              (s.maxMinerFee, s.params.baseAmount, s.params.redeemer.toErgoTree, s.params.minQuoteAmount).pure[F]
+            case s: SwapMultiAddress =>
+              Either
+                .catchNonFatal(ErgoTreeSerializer.default.deserialize(s.params.redeemer))
+                .leftMap(err => IncorrectMultiAddressTree(pool.poolId, s.box.boxId, s.params.redeemer, err.getMessage))
+                .toRaise
+                .map(tree => (s.maxMinerFee, s.params.baseAmount, tree, s.params.minQuoteAmount))
+          }).map { case (maxMinerFee, inputSwap, redeemer, minOutput) =>
+            val poolBox0 = pool.box
+            val swapBox  = swap.box
+            val swapIn   = new Input(swapBox.boxId.toErgo, ProverResult.empty)
+            val poolIn   = new Input(poolBox0.boxId.toErgo, ProverResult.empty)
+            val dexFeeIn = new Input(dexFeeOutput.boxId.toErgo, ProverResult.empty)
+            val (deltaX, deltaY) =
+              if (input.id == pool.x.id) input.value -> -output.value
+              else -output.value                     -> input.value
+            val poolBox1 = new ErgoBoxCandidate(
+              value          = poolBox0.value + deltaX,
+              ergoTree       = contracts.pool,
+              creationHeight = ctx.currentHeight,
+              additionalTokens = mkPoolTokens(
+                pool,
+                amountLP = pool.lp.value,
+                amountY  = pool.y.value + deltaY
+              ),
+              additionalRegisters = mkPoolRegs(pool)
+            )
+            val minerFee    = execution.minerFee min maxMinerFee
+            val minerFeeBox = new ErgoBoxCandidate(minerFee, minerFeeProp, ctx.currentHeight)
+            val dexFeeBox = new ErgoBoxCandidate(
+              dexFeeOutput.value + dexFee,
+              P2PKAddress(sk.publicImage).script,
+              ctx.currentHeight,
+              additionalTokens = mkTokens(dexFeeOutput.assets.map(asset => asset.tokenId -> asset.amount): _*)
+            )
+            val rewardBox =
+              if (inputSwap.isNative)
+                new ErgoBoxCandidate(
+                  value            = swapBox.value - input.value - minerFeeBox.value - dexFee,
+                  ergoTree         = redeemer,
+                  creationHeight   = ctx.currentHeight,
+                  additionalTokens = mkTokens(minOutput.id -> output.value)
+                )
```

Comment:
It's not good that we have to construct ErgoBoxes right among the business logic of execution. In Haskell impl we have `ToLedger[A]` instances for this.